esp/bindings/http/platform/httpprot.cpp

```diff
@@ -516,9 +512,18 @@ CPooledHttpThread::~CPooledHttpThread()
 
 void CPooledHttpThread::threadmain()
 {
+    // Ensure the socket is returned before leaving this method. There is no reason to retain
+    // sockets until the thread is reused.
+    struct SocketReturner
+    {
+        CPooledHttpThread& owner;
+        SocketReturner(CPooledHttpThread& _owner) : owner(_owner) {}
+        ~SocketReturner() { owner.returnSocket();}
+    } sr(*this);
```

Comment:
so what's the magic here? what is struct sr doing?

Reply:
It's forcing the socket to be returned for every exit path of the method. I had tried creating a new method to do the work of threadmain, leaving thread main to do nothing but call that method and return the socket. Valgrind still complained about leaked sockets. This was the approach that made it happy.